_source/_posts/2019-09-18-build-a-simple-dotnet-core-app-in-docker.md

```diff
@@ -0,0 +1,343 @@
+
+## Manage Dependencies Efficiently with Docker
+
+Reading this, you may be thinking, why bother to copy the project file and run `restore` before copying the source code and running `build`. Why not copy everything then build and restore in one step? The answer is caching. Every time a `Dockerfile` modifies the docker image, Docker creates a snapshot.  If you copy a file or run a command to install a package, Docker captures the differences in a new snapshot. Docker then caches and reuses the snapshots if the image hasn't changed. So, by restoring dependencies as a separate step, the image snapshot can be reused for every build, as long as the dependencies haven't changed. This process speeds up the build considerably since downloading dependencies can take some time.
+
+## Run the ASP.NET Core App in a Docker Container
+
+As mentioned above, a `Dockerfile` can be considered a series of filesystem transformations. Your current file transforms the Microsoft-provided SDK container into a new container with both the Microsoft SDK and a release build of your application stored at `/app`.
+
+Check this out::
+
+```bash
+# Build an image using the Dockerfile in the current directory
+docker build --target build -t oktamvclogin .
+# Run the image, executing the command 'ls /app'
+docker run -it oktamvclogin ls /app
+
+```
+
+You'll see that the `app` folder in your container image contains the Release build output for your project.
+
+## Remove ASP.NET Core Development Tools from Your Docker Image
+
+So far, you've built your application within a Docker container. Nice work!
+
+However, remember Docker, as a tool, reduces the number of moving parts in your application. To improve reliability by eliminating unnecessary dependencies, we also need to remove development tools, which can cause conflicts and open security risks. The Microsoft-provided SDK image includes development tools, so let’s look at how to get rid of them.
+
+Add the following lines to your `Dockerfile`:
+
+```bash
+FROM build AS publish
+RUN dotnet publish "OktaMvcLogin.csproj" -c Release -o /app
+
+FROM mcr.microsoft.com/dotnet/core/aspnet:2.2 AS base
+WORKDIR /app
+EXPOSE 5001
+
+FROM base AS final
+WORKDIR /app
+COPY --from=publish /app .
+ENTRYPOINT ["dotnet", "OktaMvcLogin.dll"]
+```
+
+You will see a few `FROM` commands, each with an `AS` clause. This syntax provides multi-stage builds, the key to getting rid of unnecessary dependencies. In plain English, your build process is now:
+
+1. Use the SDK image to create a release build of the application. Call this stage 'build'
+2. Use the 'build' stage image to publish the application to the 'app' folder. Call this stage 'publish'
+3. Download the Microsoft-provided ASP.NET core image, which has only runtime components. Call this stage 'base'
+4. Using the 'base' stage image, copy the contents of the 'app' directory from the 'publish' stage. Call this stage 'final'
```

Comment:
I would add periods at the end of each item since the previous sentence has a period.